src/System.Memory/src/System/Buffers/BuffersExtensions.cs

```diff
@@ -139,6 +139,10 @@ private static void WriteMultiSegment<T>(IBufferWriter<T> writer, in ReadOnlySpa
                 if (input.Length > 0)
                 {
                     destination = writer.GetSpan();
+
+                    if (destination.IsEmpty)
+                        ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.writer);
```

Comment:
nit: I think `ArgumentException` is more appropriate than `ArgumentOutOfRangeException`, which I typically reserve for when a fixed range of values are allowed, which never makes sense in a reference type. Here, the argument misbehaved. We don't have a great exception for that, but `ArgumentException` seems reasonable enough.
Of course that would require that you come up with an exception message to explain what went wrong since you can't just specify a paramName. That's a good thing though, since the person who experiences the exception as you have it here would very likely be quite confused for a while as to why this exception would be thrown.

Reply:
I agree that this is not perfect.

Similar methods solve the problem in two ways:

`ReadOnlySequence<T>.CopyTo(Span<T>)` throws `ArgumentOutOfRangeException`

https://github.com/dotnet/corefx/blob/1326928684c6ade5482e69e1b435234c554ef3b6/src/System.Memory/src/System/Buffers/BuffersExtensions.cs#L70

`Span<T>.CopyTo(Span<T>)` throws `ArgumentException_DestinationTooShort`

https://github.com/dotnet/corefx/blob/b2097cbdcb26f7f317252334ddcce101a20b7f3d/src/Common/src/CoreLib/System/Span.Fast.cs#L266

To be consistent with other methods from `BuffersExtensions` I am going to stay with `ArgumentOutOfRangeException` but to improve the end user experience I am going to add the information about thrown exception to the summary comment.